File: projetos-modulos/membros/backup_database.py
#!/usr/bin/env python3
"""
Script de Backup e Restore do Banco de Dados
Módulo de Cadastro de Membros - Sistema de Gestão Paroquial
"""

import mysql.connector
import os
import sys
import json
import argparse
from datetime import datetime
from typing import List, Dict, Any, Optional

# Cores para output
class Colors:
    RED = '\033[31m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    MAGENTA = '\033[35m'
    CYAN = '\033[36m'
    WHITE = '\033[37m'
    RESET = '\033[0m'

def colorize(text: str, color: str) -> str:
    """Aplica cor ao texto"""
    return f"{getattr(Colors, color.upper())}{text}{Colors.RESET}"

def print_header(title: str):
    """Imprime cabeçalho formatado"""
    print(f"\n{colorize('=' * 62, 'cyan')}")
    print(f"{colorize('  ' + title.upper(), 'cyan')}")
    print(f"{colorize('=' * 62, 'cyan')}\n")

def print_step(step: int, description: str):
    """Imprime passo do processo"""
    print(f"{colorize(f'[{step}]', 'yellow')} {description}")

def print_success(message: str):
    """Imprime mensagem de sucesso"""
    print(f"{colorize('✓', 'green')} {message}")

def print_error(message: str):
    """Imprime mensagem de erro"""
    print(f"{colorize('✗', 'red')} {message}")

def print_warning(message: str):
    """Imprime mensagem de aviso"""
    print(f"{colorize('⚠', 'yellow')} {message}")

def print_info(message: str):
    """Imprime mensagem informativa"""
    print(f"{colorize('ℹ', 'blue')} {message}")

def format_bytes(bytes_value: int, precision: int = 2) -> str:
    """Formata bytes em unidades legíveis"""
    units = ['B', 'KB', 'MB', 'GB', 'TB']
    
    for i in range(len(units)):
        if bytes_value < 1024.0 or i == len(units) - 1:
            return f"{bytes_value:.{precision}f} {units[i]}"
        bytes_value /= 1024.0

class DatabaseManager:
    """Gerenciador de banco de dados"""
    
    def __init__(self):
        self.config = {
            'host': 'gerencialparoq.mysql.dbaas.com.br',
            'database': 'gerencialparoq',
            'user': 'gerencialparoq',
            'password': 'Dsg#1806',
            'charset': 'utf8mb4',
            'autocommit': False
        }
        self.connection = None
        self.cursor = None
    
    def connect(self) -> bool:
        """Conecta ao banco de dados"""
        try:
            self.connection = mysql.connector.connect(**self.config)
            self.cursor = self.connection.cursor(dictionary=True)
            return True
        except mysql.connector.Error as e:
            print_error(f"Erro ao conectar: {e}")
            return False
    
    def disconnect(self):
        """Desconecta do banco de dados"""
        if self.cursor:
            self.cursor.close()
        if self.connection:
            self.connection.close()
    
    def execute_query(self, query: str, params: tuple = None) -> bool:
        """Executa uma query"""
        try:
            self.cursor.execute(query, params)
            self.connection.commit()
            return True
        except mysql.connector.Error as e:
            print_error(f"Erro ao executar query: {e}")
            return False
    
    def fetch_one(self, query: str, params: tuple = None) -> Dict[str, Any]:
        """Busca um resultado"""
        try:
            self.cursor.execute(query, params)
            return self.cursor.fetchone()
        except mysql.connector.Error as e:
            print_error(f"Erro ao buscar resultado: {e}")
            return None
    
    def fetch_all(self, query: str, params: tuple = None) -> List[Dict[str, Any]]:
        """Busca todos os resultados"""
        try:
            self.cursor.execute(query, params)
            return self.cursor.fetchall()
        except mysql.connector.Error as e:
            print_error(f"Erro ao buscar resultados: {e}")
            return []
    
    def test_connection(self) -> bool:
        """Testa a conexão"""
        try:
            result = self.fetch_one("SELECT 1 as test")
            return result and result['test'] == 1
        except:
            return False

class BackupManager:
    """Gerenciador de backup e restore"""
    
    def __init__(self, db_manager: DatabaseManager):
        self.db = db_manager
        self.backup_dir = os.path.join(os.path.dirname(__file__), 'backups')
        self.timestamp = datetime.now().strftime('%Y-%m-%d_%H-%M-%S')
        
        # Criar diretório de backup se não existir
        if not os.path.exists(self.backup_dir):
            os.makedirs(self.backup_dir, exist_ok=True)
    
    def get_module_tables(self) -> List[str]:
        """Obtém lista de tabelas do módulo"""
        tables = self.db.fetch_all("SHOW TABLES LIKE 'membros_%'")
        return [list(table.values())[0] for table in tables]
    
    def create_backup(self) -> Optional[str]:
        """Cria backup do módulo"""
        try:
            print_header("CRIANDO BACKUP DO MÓDULO DE MEMBROS")
            
            # Conectar ao banco
            print_step(1, "Conectando ao banco de dados...")
            if not self.db.connect():
                return None
            print_success("Conexão estabelecida com sucesso!")
            
            # Listar tabelas do módulo
            print_step(2, "Identificando tabelas do módulo...")
            tables = self.get_module_tables()
            
            if not tables:
                print_error("Nenhuma tabela do módulo encontrada!")
                return None
            
            print_success(f"Encontradas {len(tables)} tabelas do módulo")
            for table in tables:
                print(f"  - {colorize(table, 'cyan')}")
            
            # Criar arquivo de backup
            backup_file = os.path.join(self.backup_dir, f"membros_backup_{self.timestamp}.sql")
            print_step(3, f"Criando arquivo de backup: {os.path.basename(backup_file)}")
            
            backup_content = f"""-- Backup do Módulo de Membros
-- Data: {datetime.now().strftime('%d/%m/%Y %H:%M:%S')}
-- Sistema: GerencialParoq
-- Python Script

SET FOREIGN_KEY_CHECKS = 0;

"""
            
            # Backup de estrutura e dados
            for table in tables:
                print_step(4, f"Fazendo backup da tabela: {table}")
                
                # Estrutura da tabela
                create_table = self.db.fetch_one(f"SHOW CREATE TABLE `{table}`")
                if create_table:
                    backup_content += f"-- Estrutura da tabela {table}\n"
                    backup_content += f"DROP TABLE IF EXISTS `{table}`;\n"
                    backup_content += f"{create_table['Create Table']};\n\n"
                
                # Dados da tabela
                rows = self.db.fetch_all(f"SELECT * FROM `{table}`")
                if rows:
                    backup_content += f"-- Dados da tabela {table}\n"
                    
                    # Obter nomes das colunas
                    columns = list(rows[0].keys())
                    column_names = '`' + '`, `'.join(columns) + '`'
                    
                    # Inserir dados em lotes
                    batch_size = 100
                    total_rows = len(rows)
                    batches = (total_rows + batch_size - 1) // batch_size
                    
                    for i in range(batches):
                        start = i * batch_size
                        end = min(start + batch_size, total_rows)
                        batch = rows[start:end]
                        
                        backup_content += f"INSERT INTO `{table}` ({column_names}) VALUES\n"
                        
                        values = []
                        for row in batch:
                            row_values = []
                            for value in row.values():
                                if value is None:
                                    row_values.append('NULL')
                                elif isinstance(value, str):
                                    row_values.append(f"'{value.replace(chr(39), chr(39)+chr(39))}'")
                                else:
                                    row_values.append(str(value))
                            values.append('(' + ', '.join(row_values) + ')')
                        
                        backup_content += ',\n'.join(values) + ";\n\n"
                    
                    print_success(f"{table}: {len(rows)} registros")
                else:
                    print_info(f"{table}: Tabela vazia")
            
            backup_content += "SET FOREIGN_KEY_CHECKS = 1;\n\n"
            backup_content += f"-- Backup concluído em {datetime.now().strftime('%d/%m/%Y %H:%M:%S')}\n"
            
            # Salvar arquivo
            with open(backup_file, 'w', encoding='utf-8') as f:
                f.write(backup_content)
            
            file_size = os.path.getsize(backup_file)
            file_size_formatted = format_bytes(file_size)
            
            print_success("Backup criado com sucesso!")
            print_info(f"Arquivo: {os.path.basename(backup_file)}")
            print_info(f"Tamanho: {file_size_formatted}")
            print_info(f"Localização: {backup_file}")
            
            # Criar arquivo de metadados
            metadata_file = os.path.join(self.backup_dir, f"membros_backup_{self.timestamp}.json")
            metadata = {
                'timestamp': self.timestamp,
                'date': datetime.now().strftime('%d/%m/%Y %H:%M:%S'),
                'tables': tables,
                'table_count': len(tables),
                'file_size': file_size,
                'file_size_formatted': file_size_formatted,
                'version': '1.0',
                'system': 'GerencialParoq - Módulo de Membros'
            }
            
            with open(metadata_file, 'w', encoding='utf-8') as f:
                json.dump(metadata, f, indent=2, ensure_ascii=False)
            
            print_success(f"Metadados salvos: {os.path.basename(metadata_file)}")
            
            return backup_file
            
        except Exception as e:
            print_error(f"Erro ao criar backup: {e}")
            return None
        finally:
            self.db.disconnect()
    
    def restore_backup(self, backup_file: str) -> bool:
        """Restaura backup do módulo"""
        try:
            print_header("RESTAURANDO BACKUP DO MÓDULO DE MEMBROS")
            
            if not os.path.exists(backup_file):
                print_error(f"Arquivo de backup não encontrado: {backup_file}")
                return False
            
            # Conectar ao banco
            print_step(1, "Conectando ao banco de dados...")
            if not self.db.connect():
                return False
            print_success("Conexão estabelecida com sucesso!")
            
            # Ler arquivo de backup
            print_step(2, "Lendo arquivo de backup...")
            with open(backup_file, 'r', encoding='utf-8') as f:
                backup_content = f.read()
            
            file_size = os.path.getsize(backup_file)
            file_size_formatted = format_bytes(file_size)
            print_success(f"Arquivo lido: {os.path.basename(backup_file)} ({file_size_formatted})")
            
            # Executar backup
            print_step(3, "Executando restauração...")
            statements = [stmt.strip() for stmt in backup_content.split(';') if stmt.strip() and not stmt.strip().startswith('--')]
            
            success_count = 0
            error_count = 0
            
            for statement in statements:
                if self.db.execute_query(statement):
                    success_count += 1
                else:
                    error_count += 1
            
            print_success(f"Restauração concluída! ({success_count} comandos executados, {error_count} erros)")
            
            # Verificar restauração
            print_step(4, "Verificando restauração...")
            tables = self.get_module_tables()
            
            print_success(f"Tabelas restauradas: {len(tables)}")
            for table in tables:
                result = self.db.fetch_one(f"SELECT COUNT(*) as count FROM `{table}`")
                count = result['count'] if result else 0
                print(f"  - {colorize(table, 'cyan')}: {count} registros")
            
            return error_count == 0
            
        except Exception as e:
            print_error(f"Erro ao restaurar backup: {e}")
            return False
        finally:
            self.db.disconnect()
    
    def list_backups(self) -> List[Dict[str, Any]]:
        """Lista backups disponíveis"""
        print_header("LISTANDO BACKUPS DISPONÍVEIS")
        
        backup_files = []
        for file in os.listdir(self.backup_dir):
            if file.startswith('membros_backup_') and file.endswith('.sql'):
                file_path = os.path.join(self.backup_dir, file)
                file_stat = os.stat(file_path)
                
                backup_info = {
                    'filename': file,
                    'filepath': file_path,
                    'size': file_stat.st_size,
                    'size_formatted': format_bytes(file_stat.st_size),
                    'date': datetime.fromtimestamp(file_stat.st_mtime).strftime('%d/%m/%Y %H:%M:%S'),
                    'metadata': None
                }
                
                # Tentar carregar metadados
                metadata_file = file_path.replace('.sql', '.json')
                if os.path.exists(metadata_file):
                    try:
                        with open(metadata_file, 'r', encoding='utf-8') as f:
                            backup_info['metadata'] = json.load(f)
                    except:
                        pass
                
                backup_files.append(backup_info)
        
        # Ordenar por data de modificação (mais recente primeiro)
        backup_files.sort(key=lambda x: os.path.getmtime(x['filepath']), reverse=True)
        
        if not backup_files:
            print_info("Nenhum backup encontrado.")
            return []
        
        print(f"{colorize('📁', 'blue')} Backups encontrados: {len(backup_files)}\n")
        
        for i, backup in enumerate(backup_files, 1):
            print(f"{colorize(f'[{i}]', 'yellow')} {backup['filename']}")
            print(f"    Data: {backup['date']}")
            print(f"    Tamanho: {backup['size_formatted']}")
            
            if backup['metadata']:
                print(f"    Tabelas: {backup['metadata'].get('table_count', 'N/A')}")
                print(f"    Sistema: {backup['metadata'].get('system', 'N/A')}")
            print()
        
        return backup_files

def main():
    """Função principal"""
    parser = argparse.ArgumentParser(description='Backup e Restore do Módulo de Membros')
    parser.add_argument('action', choices=['backup', 'restore', 'list'], 
                       help='Ação a executar: backup, restore ou list')
    parser.add_argument('--file', '-f', help='Arquivo de backup para restore')
    
    args = parser.parse_args()
    
    # Verificar se o MySQL connector está instalado
    try:
        import mysql.connector
    except ImportError:
        print_error("MySQL Connector não está instalado!")
        print_info("Execute: pip install mysql-connector-python")
        return False
    
    db_manager = DatabaseManager()
    backup_manager = BackupManager(db_manager)
    
    if args.action == 'backup':
        return backup_manager.create_backup() is not None
    elif args.action == 'restore':
        if not args.file:
            print_error("Especifique o arquivo de backup para restaurar.")
            print_info("Uso: python backup_database.py restore --file arquivo.sql")
            return False
        
        # Se o arquivo não tem caminho completo, procurar na pasta de backups
        if not os.path.isabs(args.file):
            args.file = os.path.join(backup_manager.backup_dir, args.file)
        
        return backup_manager.restore_backup(args.file)
    elif args.action == 'list':
        backup_manager.list_backups()
        return True

if __name__ == "__main__":
    try:
        success = main()
        sys.exit(0 if success else 1)
    except KeyboardInterrupt:
        print(f"\n{colorize('⚠', 'yellow')} Operação cancelada pelo usuário.")
        sys.exit(1)
    except Exception as e:
        print(f"\n{colorize('✗', 'red')} Erro inesperado: {e}")
        sys.exit(1)

File: projetos-modulos/membros/backup_database.php
<?php
/**
 * Script de Backup e Restore do Banco de Dados
 * Módulo de Cadastro de Membros - Sistema de Gestão Paroquial
 */

// Incluir configuração do banco
require_once 'config/database.php';

// Cores para output
$colors = [
    'red' => "\033[31m",
    'green' => "\033[32m",
    'yellow' => "\033[33m",
    'blue' => "\033[34m",
    'magenta' => "\033[35m",
    'cyan' => "\033[36m",
    'white' => "\033[37m",
    'reset' => "\033[0m"
];

function colorize($text, $color) {
    global $colors;
    return $colors[$color] . $text . $colors['reset'];
}

function printHeader($title) {
    echo "\n" . colorize("=" . str_repeat("=", 60) . "=", 'cyan') . "\n";
    echo colorize("  " . strtoupper($title), 'cyan') . "\n";
    echo colorize("=" . str_repeat("=", 60) . "=", 'cyan') . "\n\n";
}

function printStep($step, $description) {
    echo colorize("[$step] ", 'yellow') . $description . "\n";
}

function printSuccess($message) {
    echo colorize("✓ ", 'green') . $message . "\n";
}

function printError($message) {
    echo colorize("✗ ", 'red') . $message . "\n";
}

function printInfo($message) {
    echo colorize("ℹ ", 'blue') . $message . "\n";
}

// Configurações
$backupDir = __DIR__ . '/backups';
$timestamp = date('Y-m-d_H-i-s');

// Criar diretório de backup se não existir
if (!is_dir($backupDir)) {
    mkdir($backupDir, 0755, true);
}

// Função para fazer backup
function createBackup($backupDir, $timestamp) {
    try {
        printHeader("CRIANDO BACKUP DO MÓDULO DE MEMBROS");
        
        // Conectar ao banco
        printStep("1", "Conectando ao banco de dados...");
        $db = new MembrosDatabase();
        
        if (!$db->testConnection()) {
            throw new Exception("Falha na conexão com o banco de dados");
        }
        printSuccess("Conexão estabelecida com sucesso!");
        
        // Listar tabelas do módulo
        printStep("2", "Identificando tabelas do módulo...");
        $tables = $db->fetchAll("SHOW TABLES LIKE 'membros_%'");
        $tableNames = array_column($tables, 'Tables_in_gerencialparoq (membros_%)');
        
        if (empty($tableNames)) {
            throw new Exception("Nenhuma tabela do módulo encontrada!");
        }
        
        printSuccess("Encontradas " . count($tableNames) . " tabelas do módulo");
        foreach ($tableNames as $table) {
            echo "  - " . colorize($table, 'cyan') . "\n";
        }
        
        // Criar arquivo de backup
        $backupFile = $backupDir . "/membros_backup_$timestamp.sql";
        printStep("3", "Criando arquivo de backup: " . basename($backupFile));
        
        $backupContent = "-- Backup do Módulo de Membros\n";
        $backupContent .= "-- Data: " . date('d/m/Y H:i:s') . "\n";
        $backupContent .= "-- Sistema: GerencialParoq\n\n";
        
        $backupContent .= "SET FOREIGN_KEY_CHECKS = 0;\n\n";
        
        // Backup de estrutura e dados
        foreach ($tableNames as $table) {
            printStep("4", "Fazendo backup da tabela: $table");
            
            // Estrutura da tabela
            $createTable = $db->fetchOne("SHOW CREATE TABLE `$table`");
            $backupContent .= "-- Estrutura da tabela $table\n";
            $backupContent .= "DROP TABLE IF EXISTS `$table`;\n";
            $backupContent .= $createTable['Create Table'] . ";\n\n";
            
            // Dados da tabela
            $rows = $db->fetchAll("SELECT * FROM `$table`");
            if (!empty($rows)) {
                $backupContent .= "-- Dados da tabela $table\n";
                
                // Obter nomes das colunas
                $columns = array_keys($rows[0]);
                $columnNames = '`' . implode('`, `', $columns) . '`';
                
                // Inserir dados em lotes
                $batchSize = 100;
                $totalRows = count($rows);
                $batches = ceil($totalRows / $batchSize);
                
                for ($i = 0; $i < $batches; $i++) {
                    $start = $i * $batchSize;
                    $end = min($start + $batchSize, $totalRows);
                    $batch = array_slice($rows, $start, $end - $start);
                    
                    $backupContent .= "INSERT INTO `$table` ($columnNames) VALUES\n";
                    
                    $values = [];
                    foreach ($batch as $row) {
                        $rowValues = [];
                        foreach ($row as $value) {
                            if ($value === null) {
                                $rowValues[] = 'NULL';
                            } else {
                                $rowValues[] = "'" . addslashes($value) . "'";
                            }
                        }
                        $values[] = '(' . implode(', ', $rowValues) . ')';
                    }
                    
                    $backupContent .= implode(",\n", $values) . ";\n\n";
                }
                
                printSuccess("$table: " . count($rows) . " registros");
            } else {
                printInfo("$table: Tabela vazia");
            }
        }
        
        $backupContent .= "SET FOREIGN_KEY_CHECKS = 1;\n\n";
        $backupContent .= "-- Backup concluído em " . date('d/m/Y H:i:s') . "\n";
        
        // Salvar arquivo
        if (file_put_contents($backupFile, $backupContent) === false) {
            throw new Exception("Erro ao salvar arquivo de backup");
        }
        
        $fileSize = filesize($backupFile);
        $fileSizeFormatted = formatBytes($fileSize);
        
        printSuccess("Backup criado com sucesso!");
        printInfo("Arquivo: " . basename($backupFile));
        printInfo("Tamanho: $fileSizeFormatted");
        printInfo("Localização: $backupFile");
        
        // Criar arquivo de metadados
        $metadataFile = $backupDir . "/membros_backup_$timestamp.json";
        $metadata = [
            'timestamp' => $timestamp,
            'date' => date('d/m/Y H:i:s'),
            'tables' => $tableNames,
            'table_count' => count($tableNames),
            'file_size' => $fileSize,
            'file_size_formatted' => $fileSizeFormatted,
            'version' => '1.0',
            'system' => 'GerencialParoq - Módulo de Membros'
        ];
        
        file_put_contents($metadataFile, json_encode($metadata, JSON_PRETTY_PRINT));
        printSuccess("Metadados salvos: " . basename($metadataFile));
        
        return $backupFile;
        
    } catch (Exception $e) {
        printError("Erro ao criar backup: " . $e->getMessage());
        return false;
    }
}

// Função para restaurar backup
function restoreBackup($backupFile) {
    try {
        printHeader("RESTAURANDO BACKUP DO MÓDULO DE MEMBROS");
        
        if (!file_exists($backupFile)) {
            throw new Exception("Arquivo de backup não encontrado: $backupFile");
        }
        
        // Conectar ao banco
        printStep("1", "Conectando ao banco de dados...");
        $db = new MembrosDatabase();
        
        if (!$db->testConnection()) {
            throw new Exception("Falha na conexão com o banco de dados");
        }
        printSuccess("Conexão estabelecida com sucesso!");
        
        // Ler arquivo de backup
        printStep("2", "Lendo arquivo de backup...");
        $backupContent = file_get_contents($backupFile);
        
        if ($backupContent === false) {
            throw new Exception("Erro ao ler arquivo de backup");
        }
        
        $fileSize = filesize($backupFile);
        $fileSizeFormatted = formatBytes($fileSize);
        printSuccess("Arquivo lido: " . basename($backupFile) . " ($fileSizeFormatted)");
        
        // Executar backup
        printStep("3", "Executando restauração...");
        $statements = explode(';', $backupContent);
        
        $successCount = 0;
        $errorCount = 0;
        
        foreach ($statements as $statement) {
            $statement = trim($statement);
            if (empty($statement) || strpos($statement, '--') === 0) {
                continue;
            }
            
            try {
                $db->execute($statement);
                $successCount++;
            } catch (Exception $e) {
                $errorCount++;
                printError("Erro ao executar: " . substr($statement, 0, 50) . "... - " . $e->getMessage());
            }
        }
        
        printSuccess("Restauração concluída! ($successCount comandos executados, $errorCount erros)");
        
        // Verificar restauração
        printStep("4", "Verificando restauração...");
        $tables = $db->fetchAll("SHOW TABLES LIKE 'membros_%'");
        $tableNames = array_column($tables, 'Tables_in_gerencialparoq (membros_%)');
        
        printSuccess("Tabelas restauradas: " . count($tableNames));
        foreach ($tableNames as $table) {
            $count = $db->fetchOne("SELECT COUNT(*) as count FROM `$table`")['count'];
            echo "  - " . colorize("$table", 'cyan') . ": $count registros\n";
        }
        
        return true;
        
    } catch (Exception $e) {
        printError("Erro ao restaurar backup: " . $e->getMessage());
        return false;
    }
}

// Função para listar backups
function listBackups($backupDir) {
    printHeader("LISTANDO BACKUPS DISPONÍVEIS");
    
    $backups = glob($backupDir . "/membros_backup_*.sql");
    
    if (empty($backups)) {
        printInfo("Nenhum backup encontrado.");
        return;
    }
    
    // Ordenar por data de modificação (mais recente primeiro)
    usort($backups, function($a, $b) {
        return filemtime($b) - filemtime($a);
    });
    
    echo colorize("📁 ", 'blue') . "Backups encontrados: " . count($backups) . "\n\n";
    
    foreach ($backups as $i => $backup) {
        $filename = basename($backup);
        $fileSize = formatBytes(filesize($backup));
        $fileDate = date('d/m/Y H:i:s', filemtime($backup));
        
        echo colorize("[" . ($i + 1) . "] ", 'yellow') . "$filename\n";
        echo "    Data: $fileDate\n";
        echo "    Tamanho: $fileSize\n";
        
        // Verificar se existe arquivo de metadados
        $metadataFile = str_replace('.sql', '.json', $backup);
        if (file_exists($metadataFile)) {
            $metadata = json_decode(file_get_contents($metadataFile), true);
            if ($metadata) {
                echo "    Tabelas: " . $metadata['table_count'] . "\n";
                echo "    Sistema: " . $metadata['system'] . "\n";
            }
        }
        echo "\n";
    }
}

// Função para formatar bytes
function formatBytes($bytes, $precision = 2) {
    $units = array('B', 'KB', 'MB', 'GB', 'TB');
    
    for ($i = 0; $bytes > 1024 && $i < count($units) - 1; $i++) {
        $bytes /= 1024;
    }
    
    return round($bytes, $precision) . ' ' . $units[$i];
}

// Menu principal
function showMenu() {
    echo "\n" . colorize("🔧 ", 'cyan') . "MENU DE BACKUP E RESTORE\n";
    echo "1. Criar backup\n";
    echo "2. Restaurar backup\n";
    echo "3. Listar backups\n";
    echo "4. Sair\n\n";
}

// Execução principal
if (php_sapi_name() === 'cli') {
    // Modo CLI
    $action = $argv[1] ?? 'menu';
    
    switch ($action) {
        case 'backup':
            createBackup($backupDir, $timestamp);
            break;
            
        case 'restore':
            if (isset($argv[2])) {
                $backupFile = $backupDir . '/' . $argv[2];
                restoreBackup($backupFile);
            } else {
                printError("Especifique o arquivo de backup para restaurar.");
                echo "Uso: php backup_database.php restore arquivo.sql\n";
            }
            break;
            
        case 'list':
            listBackups($backupDir);
            break;
            
        case 'menu':
        default:
            showMenu();
            $choice = readline("Escolha uma opção (1-4): ");
            
            switch ($choice) {
                case '1':
                    createBackup($backupDir, $timestamp);
                    break;
                case '2':
                    listBackups($backupDir);
                    $backupFile = readline("Digite o nome do arquivo para restaurar: ");
                    if ($backupFile) {
                        restoreBackup($backupDir . '/' . $backupFile);
                    }
                    break;
                case '3':
                    listBackups($backupDir);
                    break;
                case '4':
                    echo "Saindo...\n";
                    exit(0);
                default:
                    printError("Opção inválida!");
            }
            break;
    }
} else {
    // Modo web
    printHeader("BACKUP E RESTORE - MÓDULO DE MEMBROS");
    echo "Este script deve ser executado via linha de comando.\n";
    echo "Uso: php backup_database.php [backup|restore|list]\n";
}
?>

